### PR TITLE
Fix namespace set on upgrade command

### PR DESCRIPTION
**Pull request by Itxaka**
Namespace was not being set via annotations properly
for the upgrade command nor the install command called
from the upgrade

Fixes: #73 

Signed-off-by: Itxaka <igarcia@suse.com>

### CODE REVIEW

File: cmd/hypper/upgrade.go [hunk 1]
```diff
@@ -114,6 +122,12 @@ func newUpgradeCmd(cfg *action.Configuration, logger log.Logger) *cobra.Command
 						logger.Info(eyecandy.ESPrintf(settings.NoEmojis, "Release %q does not exist. Installing it now. :toolbox:\n", client.ReleaseName))
 					}
 					instClient := action.NewInstall(cfg)
+					// Set namespace for the install client
```
Comment (dragonchaser) [inline cmd/hypper/upgrade.go:125]: we have almost identical code in three places now (twice here line 125-130, 168-174 and in install.go) can we move that to a static helper function?

File: cmd/hypper/upgrade.go [hunk 2]
```diff
@@ -151,6 +165,13 @@ func newUpgradeCmd(cfg *action.Configuration, logger log.Logger) *cobra.Command
 				return err
 			}
 
+			// Set namespace for the upgrade client
```
Comment (dragonchaser) [inline cmd/hypper/upgrade.go:168]: see previous comment

**Review by dragonchaser — CHANGES_REQUESTED** [review]

File: cmd/hypper/upgrade.go (hunk 1 above)
Comment (viccuad) [reply cmd/hypper/upgrade.go:125]: I agree. I would schedule that inside https://github.com/rancher-sandbox/hypper/issues/70 (and rename the card to "fix name and namespace setting deduplication" or so).

Comment (viccuad) [reply cmd/hypper/upgrade.go:125]: I would prefer to see all duplications everywhere, to see the pattern, and then fix them in a way that is good for all callers, hence why I want to postpone that.

File: cmd/hypper/upgrade.go [hunk 3]
```diff
@@ -42,6 +42,14 @@ argument can be either: a chart reference('example/mariadb'), a path to a chart
 a packaged chart, or a fully qualified URL. For chart references, the latest
 version will be specified unless the '--version' flag is set.
 
+There are four different ways you can select the release name and namespace
+where the chart will be installed. By priority order:
```
Comment (viccuad) [inline cmd/hypper/upgrade.go:46]:
```suggestion
where the chart will be upgraded. By priority order:
```

File: pkg/action/upgrade.go [hunk 4]
```diff
@@ -57,3 +57,19 @@ func (i *Upgrade) Name(chart *chart.Chart, args []string) (string, error) {
 	// If we dont have our annotations then return the base name
 	return chart.Metadata.Name, nil
 }
+
+// SetNamespace sets the Namespace that should be used in action.Install
```
Comment (viccuad) [inline pkg/action/upgrade.go:61]:
```suggestion
// SetNamespace sets the Namespace that should be used in action.Upgrade
```

File: cmd/hypper/upgrade.go [hunk 5]
```diff
@@ -151,6 +165,13 @@ func newUpgradeCmd(cfg *action.Configuration, logger log.Logger) *cobra.Command
 				return err
 			}
 
+			// Set namespace for the upgrade client
+			if settings.NamespaceFromFlag {
+				client.Namespace = settings.Namespace()
+			} else {
+				client.SetNamespace(ch, settings.Namespace())
+			}
```
Comment (viccuad) [inline cmd/hypper/upgrade.go:173]: Same here, aren't we missing setting the Config namespace with a call to `client.Config.SetNamespace(client.Namespace)`?
Analogous to: https://github.com/rancher-sandbox/hypper/blob/1a686aa3a18ebb10bbb2727baa28ea7e416ccc1a/cmd/hypper/install.go#L126

File: cmd/hypper/upgrade.go [hunk 6]
```diff
@@ -114,6 +122,12 @@ func newUpgradeCmd(cfg *action.Configuration, logger log.Logger) *cobra.Command
 						logger.Info(eyecandy.ESPrintf(settings.NoEmojis, "Release %q does not exist. Installing it now. :toolbox:\n", client.ReleaseName))
 					}
 					instClient := action.NewInstall(cfg)
+					// Set namespace for the install client
+					if settings.NamespaceFromFlag {
+						instClient.Namespace = settings.Namespace()
+					} else {
+						instClient.SetNamespace(ch, settings.Namespace())
+					}
```
Comment (viccuad) [inline cmd/hypper/upgrade.go:130]: Aren't we missing setting the Config namespace with a call to `client.Config.SetNamespace(client.Namespace)`?
Analogous to: https://github.com/rancher-sandbox/hypper/blob/1a686aa3a18ebb10bbb2727baa28ea7e416ccc1a/cmd/hypper/install.go#L126

**Review by viccuad — CHANGES_REQUESTED** [review]

File: cmd/hypper/upgrade.go (hunk 5 above)
Comment (Itxaka) [reply cmd/hypper/upgrade.go:173]: yup yup!

File: cmd/hypper/upgrade.go (hunk 1 above)
Comment (Itxaka) [reply cmd/hypper/upgrade.go:125]: Agreed, card updated!

File: cmd/hypper/upgrade.go (hunk 3 above)
Comment (Itxaka) [reply cmd/hypper/upgrade.go:46]: :man_facepalming:

File: pkg/action/upgrade.go (hunk 4 above)
Comment (Itxaka) [reply pkg/action/upgrade.go:61]: :man_facepalming: :clown_face:

File: cmd/hypper/upgrade.go (hunk 1 above)
Comment (dragonchaser) [reply cmd/hypper/upgrade.go:125]: > I would prefer to see all duplications everywhere, to see the pattern, and then fix them in a way that is good for all callers, hence why I want to postpone that.

Well, why wait? I suggest using a helper package with static functions that can be invoked wherever needed... That can't hurt. If we see a pattern we can still move that function later. But we have it in place now.

File: cmd/hypper/upgrade.go (hunk 6 above)
Comment (mattfarina) [reply cmd/hypper/upgrade.go:130]: I think the lack of this line is causing the bug I found.

File: cmd/hypper/upgrade.go (hunk 1 above)
Comment (Itxaka) [reply cmd/hypper/upgrade.go:125]: I think this is out of scope for this, as this deals with a bug which needs to be fixed, not an improvement.

Comment (viccuad) [reply cmd/hypper/upgrade.go:125]: I'm personally happy with the solution for now, we have the issue open to not forget about it.

File: pkg/action/upgrade.go [hunk 7]
```diff
@@ -57,3 +57,19 @@ func (i *Upgrade) Name(chart *chart.Chart, args []string) (string, error) {
 	// If we dont have our annotations then return the base name
 	return chart.Metadata.Name, nil
 }
+
+// SetNamespace sets the Namespace that should be used in action.Upgrade
+//
+// This will read the chart annotations. If no annotations, it leave the existing ns in the action.
+func (i *Upgrade) SetNamespace(chart *chart.Chart, defaultns string) {
```
Comment (viccuad) [inline pkg/action/upgrade.go:64]: We should add unit tests for this new function. E.g: https://github.com/rancher-sandbox/hypper/blob/viccuad%2Fadd-shared-deps/pkg/action/dependency_shared_test.go#L81-L101. The different TestSetNamespace() functions will clash, so either just rename the TestSetNamespace() to TestUpgradeSetNamespace(), or rename the others (as done here, but then I may need to rebase https://github.com/rancher-sandbox/hypper/pull/80/commits/a0acead55d686354d1a38ce5d9eb7128b8bbc46a#diff-b97a6f3723149f75e7945127b215944ea6a752bcf41a2e1c00b65ae0a409bc7cR34-R38).

**Review by viccuad — DISMISSED** [review]

File: pkg/action/upgrade.go (hunk 7 above)
Comment (Itxaka) [reply pkg/action/upgrade.go:64]: good point, didnt even think about the name clash :o

Comment (Itxaka) [reply pkg/action/upgrade.go:64]: unfortunately as I want to reuse some of your added functions I will have to wait for your Pr to merge :D

File: cmd/hypper/hypper_test.go [hunk 8]
```diff
@@ -43,7 +43,7 @@ import (
 func testTimestamper() time.Time { return time.Unix(242085845, 0).UTC() }
 
 func init() {
-	action.Timestamper = testTimestamper
+	helmAction.Timestamper = testTimestamper
```
Comment (viccuad) [inline cmd/hypper/hypper_test.go:46]: :100:  top!

**Comment by Itxaka** [comment]
Either this or it wazs already broken but passing the `release-name` flag does not work

**Comment by Itxaka** [comment]
My mistake, seems that --release-name works, it was just missing tests...adding those...

File: pkg/action/upgrade.go (hunk 7 above)
Comment (Itxaka) [reply pkg/action/upgrade.go:58]: done!

File: cmd/hypper/upgrade.go (hunk 1 above)
Comment (dragonchaser) [reply cmd/hypper/upgrade.go:125]: Ok, then we should as this as new issue.

**Review by dragonchaser — APPROVED** [review]
lgtm

File: cmd/hypper/upgrade.go (hunk 1 above)
Comment (Itxaka) [reply cmd/hypper/upgrade.go:125]: got a PR already for this :D

**Pull request merged**